device/pcie/pci_device.cpp

```diff
@@ -0,0 +1,349 @@
+        }
+
+        // Using Write-Combine memory mode. This is used for accessing DRAM on Blackhole.
+        // WC doesn't guarantee write ordering but has better performance.
+        bar4_wc_size = bar4_wc_mapping.mapping_size;
+        bar4_wc = mmap(NULL, bar4_wc_mapping.mapping_size, PROT_READ | PROT_WRITE, MAP_SHARED, device_fd, bar4_wc_mapping.mapping_base);
+
+        if (bar4_wc == MAP_FAILED) {
+            throw std::runtime_error(std::string("BAR4 WC memory mapping failed for device ") + std::to_string(device_id) + ".");
+        }
+    }
+
+    // GS+WH: ARC_SCRATCH[6], BH: NOC NODE_ID
+    read_checking_offset = arch == tt::ARCH::BLACKHOLE ? BH_NOC_NODE_ID_OFFSET : GS_WH_ARC_SCRATCH_6_OFFSET;
+}
+
+void PCIDevice::close_device() {
+    if (arch == tt::ARCH::BLACKHOLE && bar2_uc != nullptr && bar2_uc != MAP_FAILED) {
+        // Disable ATU index 0
+        // TODO: Implement disabling for all indexes, once more host channels are enabled.
+        uint64_t iatu_index = 0;
+        uint64_t iatu_base = UNROLL_ATU_OFFSET_BAR + iatu_index * 0x200;
+        uint32_t region_ctrl_2 = 0 << 31; // REGION_EN = 0
+        write_regs(reinterpret_cast<std::uint32_t*>(static_cast<uint8_t*>(bar2_uc) + iatu_base + 0x04), &region_ctrl_2, 1);
+    }
+
+    if (device_fd != -1) {
+        ::close(device_fd);
+    }
+
+    if (bar0_wc != nullptr && bar0_wc != MAP_FAILED && bar0_wc != bar0_uc) {
+        munmap(bar0_wc, bar0_wc_size);
+    }
+
+    if (bar0_uc != nullptr && bar0_uc != MAP_FAILED) {
+        munmap(bar0_uc, bar0_uc_size);
+    }
+
+    if (bar2_uc != nullptr && bar2_uc != MAP_FAILED) {
+        munmap(bar2_uc, bar2_uc_size);
+    }
+
+    if (bar4_wc != nullptr && bar4_wc != MAP_FAILED) {
+        munmap(bar4_wc, bar4_wc_size);
+    }
+
+    if (system_reg_mapping != nullptr && system_reg_mapping != MAP_FAILED) {
+        munmap(system_reg_mapping, system_reg_mapping_size);
+    }
+
+    device_fd = -1;
+    bar0_uc = nullptr;
+    bar0_wc = nullptr;
+    bar2_uc = nullptr;
+    bar4_wc = nullptr;
+    system_reg_mapping = nullptr;
+}
+
+// Open a unique device_id per host memory channel (workaround for ttkmd < 1.21 support for more than 1 pin per fd)
+void PCIDevice::open_hugepage_per_host_mem_ch(uint32_t num_host_mem_channels) {
+    for (int ch = 0; ch < num_host_mem_channels; ch++) {
+        log_debug(LogSiliconDriver, "Opening device_fd_per_host_ch device index: {} ch: {} (num_host_mem_channels: {})", device_id, ch, num_host_mem_channels);
+        int device_fd_for_host_mem = find_device(device_id);
+        if (device_fd_for_host_mem == -1) {
+            throw std::runtime_error(std::string("Failed opening a host memory device handle for device ") + std::to_string(device_id));
+        }
+        device_fd_per_host_ch.push_back(device_fd_for_host_mem);
+    }
+}
+
+tt::ARCH PCIDevice::get_arch() const {
+    return arch;
+}
+
+template<typename T>
+T* PCIDevice::get_register_address(std::uint32_t register_offset) {
```

Comment:
`volatile T*`?

Reply:
The reason I didn't make this volatile is bc `write_block` and `read_block` are using it and then using resulting address in `memcpy`: 

```
src = get_register_address<uint8_t>(byte_addr);
memcpy(dest, src, num_bytes);
```

I don't think memcpy supports taking a volatile pointer, to do so I thikn we need a workaround like in `PCIDevice::read_regs` where they did:
```
uint32_t temp = *src++;
memcpy(dest++, &temp, sizeof(temp));```

Not sure if there's confusion where volatile is needed, not sure if its' needed for `write_block` or `read_block`?
Every other areas where `get_register_address` is used will alwyas assign the output to a volatile variable